### PR TITLE
[11.0] [ADD] mdc_workstation_cards_employee_rel

### PR DESCRIPTION
**Pull request by lmiguens-solvos**
Add employee to workstation cards.
To able to put employee in a workstation card.
And when a employee changes his workstation then changes the workstation of all workstation cards  of this employee with the new workstation of the employee.

### CODE REVIEW

File: mdc_workstation_cards_employee_rel/hooks.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+        for reg in current_employees:
+            # Modify workstation cards with the employee in this workstation
+            workstation_cards = env['mdc.card'].search(
+                [('card_categ_id', '=', env.ref('mdc.mdc_card_categ_L').id)
+                    , ('workstation_id', '=', reg.id)])
+            if workstation_cards:
+                for card in workstation_cards:
+                    card.write({
+                        'employee_id': reg.current_employee_id.id
+                    })
```
Comment (dalonsod) [inline mdc_workstation_cards_employee_rel/hooks.py]: This also should work (https://www.odoo.com/documentation/11.0/reference/orm.html#odoo.models.Model.write):
```suggestion
                workstation_cards.write({
                    'employee_id': reg.current_employee_id.id
                })
```

File: mdc_workstation_cards_employee_rel/models/structure.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+class Workstation(models.Model):
+    _inherit = 'mdc.workstation'
+
+    @api.multi
+    def write(self, values):
+        self.ensure_one()
+        # Modify cards for employee changes
+        new_employee = None
+        card_categ_L = self.env.ref('mdc.mdc_card_categ_L').id
+        if 'current_employee_id' in values and values['current_employee_id']:
```
Comment (dalonsod) [inline mdc_workstation_cards_employee_rel/models/structure.py]: Code suggestion:
```suggestion
        if values.get('current_employee_id', False):
```

File: mdc_workstation_cards_employee_rel/models/structure.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                and (new_employee is None
+                     or self.current_employee_id.id != new_employee.id):
+            old_employee_cards = self.env['mdc.card'].search(
+                [('card_categ_id', '=', card_categ_L)
+                    , ('employee_id', '=', self.current_employee_id.id)])
+            if old_employee_cards:
+                for card in old_employee_cards:
+                    card.write({
+                        'workstation_id': False
+                    })
```
Comment (dalonsod) [inline mdc_workstation_cards_employee_rel/models/structure.py]: Suggested:
```suggestion
                old_employee_cards.write({
                    'workstation_id': False
                })
```

File: mdc_workstation_cards_employee_rel/models/structure.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+                and (not self.current_employee_id.id
+                     or self.current_employee_id.id != new_employee.id):
+            new_employee_cards = self.env['mdc.card'].search(
+                [('card_categ_id', '=', card_categ_L)
+                    , ('employee_id', '=', new_employee.id)])
+            if new_employee_cards:
+                for card in new_employee_cards:
+                    card.write({
+                        'workstation_id': self.id
+                    })
```
Comment (dalonsod) [inline mdc_workstation_cards_employee_rel/models/structure.py]: Suggested:
```suggestion
                new_employee_cards.write({
                    'workstation_id': self.id
                })
```

**Review by dalonsod — CHANGES_REQUESTED** [review]
Functional and code review, tested:
- Unset the employee of a workstation => workstation cards of the employee are workstation unset, but remains with the employee (OK)
- Set a unassigned workstation to this employee => the workstation cards of the employee are properly reassigned to the new workstation (OK).
- Make the same operations, but making an employee substitution => the workstation cards of the remained assigned employee are re-assigned, and the original employee are unassigned (OK).
- Emulating input and output, no problems detected (attending https://github.com/solvosci/manufacture/blob/8732facc8e95aa9e79f3e89d5ee3e03a147eb0a7/mdc/models/chkpointdata.py#L317) (OK)

Anyway, I have one doubt:
- if an employee is unassigned from a workstation, their workstation cards will be still assigned to him, permanently. Is there any situation that we'd interested in completely unassign them? If needed, then we should provide a server action for "massive unassign workstation cards only assigned to an employee" to unset them, or the user should make this work one by one. What do you think?

**Comment by lmiguens-solvos** [comment]
> Functional and code review, tested:
> 
> * Unset the employee of a workstation => workstation cards of the employee are workstation unset, but remains with the employee (OK)
> * Set a unassigned workstation to this employee => the workstation cards of the employee are properly reassigned to the new workstation (OK).
> * Make the same operations, but making an employee substitution => the workstation cards of the remained assigned employee are re-assigned, and the original employee are unassigned (OK).
> * Emulating input and output, no problems detected (attending https://github.com/solvosci/manufacture/blob/8732facc8e95aa9e79f3e89d5ee3e03a147eb0a7/mdc/models/chkpointdata.py#L317
>   ) (OK)
> 
> Anyway, I have one doubt:
> 
> * if an employee is unassigned from a workstation, their workstation cards will be still assigned to him, permanently. Is there any situation that we'd interested in completely unassign them? If needed, then we should provide a server action for "massive unassign workstation cards only assigned to an employee" to unset them, or the user should make this work one by one. What do you think?

Thanks for writing all the OK tests.

About the doubt, I understand that there may be cases in which they need to change some data (employee or position) of cards in a massive way, and that it would be good to have a massive server action, but that improvement I understand that it is not included in this budget. If in next days or months, it is necessary, it can still be attacked in the remaining hours of the bag or if this bag of hours increases.

File: mdc_workstation_cards_employee_rel/hooks.py (hunk 1 above)
Comment (lmiguens-solvos) [reply mdc_workstation_cards_employee_rel/hooks.py]: Yes, I know it. But in this case, in this model, the write method has a self.ensure_one(). Then that improvement does not work. 
https://github.com/solvosci/manufacture/blob/439c0843f25dbea63d6cddf7e7a0f9aa2842c4f4/mdc/models/structure.py#L253

File: mdc_workstation_cards_employee_rel/models/structure.py (hunk 2 above)
Comment (lmiguens-solvos) [reply mdc_workstation_cards_employee_rel/models/structure.py]: ok, I'll do it

File: mdc_workstation_cards_employee_rel/models/structure.py (hunk 3 above)
Comment (lmiguens-solvos) [reply mdc_workstation_cards_employee_rel/models/structure.py]: As i said before, in this model, the write method has a self.ensure_one()

File: mdc_workstation_cards_employee_rel/models/structure.py (hunk 4 above)
Comment (lmiguens-solvos) [reply mdc_workstation_cards_employee_rel/models/structure.py]: As i said before, in this model, the write method has a self.ensure_one()

File: mdc_workstation_cards_employee_rel/hooks.py (hunk 1 above)
Comment (dalonsod) [reply mdc_workstation_cards_employee_rel/hooks.py]: Oopps I didn't see it

**Pull request closed without merging**